Firmware/lcd.cpp

```diff
@@ -329,13 +331,30 @@ void lcd_no_autoscroll(void)
 }
 #endif
 
+/// @brief set the current LCD row
+/// @param row LCD row number, ranges from 0 to LCD_HEIGHT - 1
+static void FORCE_INLINE lcd_set_current_row(uint8_t row)
+{
+	lcd_currline = min(row, LCD_HEIGHT - 1);
+}
+
+/// @brief Calculate the LCD row offset
+/// @param row LCD row number, ranges from 0 to LCD_HEIGHT - 1
+/// @return row offset which the LCD register understands
+static uint8_t __attribute__((noinline)) lcd_get_row_offset(uint8_t row)
+{
+	return pgm_read_byte(row_offsets + min(row, LCD_HEIGHT - 1));
+}
+
 void lcd_set_cursor(uint8_t col, uint8_t row)
 {
-	uint8_t row_offsets[] = { 0x00, 0x40, 0x14, 0x54 };
-	if (row >= LCD_HEIGHT)
-		row = LCD_HEIGHT - 1;    // we count rows starting w/0
-	lcd_currline = row;  
-	lcd_command(LCD_SETDDRAMADDR | (col + row_offsets[row]));
+	lcd_set_current_row(row);
+	lcd_command(LCD_SETDDRAMADDR | (col + lcd_get_row_offset(lcd_currline)));
+}
+
+void lcd_set_cursor_column(uint8_t col)
+{
+	lcd_command(LCD_SETDDRAMADDR | (col + lcd_get_row_offset(lcd_currline)));
```

Comment:
`lcd_command(LCD_SETDDRAMADDR | (col + lcd_get_row_offset(lcd_currline)));`
seems to be duplicated in these functions - I'm not sure if doing something like this actually saves anything:
```
void lcd_set_cursor(uint8_t col, uint8_t row)
{
	lcd_set_current_row(row);
	lcd_set_cursor_column(col);
}

void lcd_set_cursor_column(uint8_t col)
{
	lcd_command(LCD_SETDDRAMADDR | (col + lcd_get_row_offset(lcd_currline)));
}
```

Reply:
yes, it saves another 22B, but `lcd_set_cursor` must be made `__attribute__((noinline))`

Reply:
IMO I would keep this as is for speed. 22B doesnt seem worth it to make the cursor change 2x slower. 

Maybe we can shorten the command to only modify the column. I need to find the datasheet again 🤔

Reply:
I would definitely keep the LCD faster at the cost of 22B. It's already painfully slow. Let's not make it even slower.
The `LCD_SETDDRAMADDR` sets both the row and the column at the same time. Actually, the LCD doesn't really have the concept of rows and columns and that's why we need the row offsets. From the perspective of the LCD, the DDRAM is just a block of memory that we write to and `LCD_SETDDRAMADDR` just updates the r/w pointer to that memory block.

Reply:
Ok, what you wrote makes sense, it would be nice to add some kind of explanation why is the code written as it is. Otherwise, I like it :wink: